navigator/backend/src/main/scala/com/digitalasset/navigator/query/project/package.scala

```diff
@@ -48,40 +48,46 @@ object project {
       rootParam: DamlLfType,
       cursor: PropertyCursor,
       value: String,
-      ps: DamlLfTypeLookup): Either[DotNotFailure, ProjectValue] = {
+      ps: DamlLfTypeLookup,
+  ): Either[DotNotFailure, ProjectValue] = {
 
     @annotation.tailrec
     def loop(
         parameter: DamlLfType,
         cursor: PropertyCursor,
-        ps: DamlLfTypeLookup): Either[DotNotFailure, ProjectValue] =
+        ps: DamlLfTypeLookup,
+    ): Either[DotNotFailure, ProjectValue] =
       parameter match {
         case tc: DamlLfTypeCon =>
-          val next = for {
-            ddt <- ps(tc.name.identifier)
-              .toRight(UnknownType(tc.name.identifier.toString, cursor, value))
-            nextCursor <- cursor.next.toRight(MustNotBeLastPart("DataType", cursor, value))
-            nextField <- tc.instantiate(ddt) match {
-              case DamlLfRecord(fields) =>
-                fields
-                  .find(f => f._1 == nextCursor.current)
-                  .toRight(UnknownProperty("record", nextCursor, value))
-              case DamlLfVariant(fields) =>
-                fields
-                  .find(f => f._1 == nextCursor.current)
-                  .toRight(UnknownProperty("variant", nextCursor, value))
-              case DamlLfEnum(_) =>
-                // FixMe (RH) https://github.com/digital-asset/daml/issues/105
-                throw new NotImplementedError("Enum types not supported")
-            }
-          } yield {
-            (nextField._2, nextCursor)
-          }
-          next match {
-            case Right((nextType, nextCursor)) => loop(nextType, nextCursor, ps)
-            case Left(e) => Left(e)
+          val tyCon = tc.name.identifier
+          ps(tyCon) match {
+            case Some(ddt) =>
+              val next = tc.instantiate(ddt) match {
+                case DamlLfEnum(_) =>
+                  Right(StringValue(tyCon.toString))
```

Comment:
Shouldn't we add a test for this?

Reply:
Ideally, yes. The code from this file is tested by [ContractSortSpec.scala](https://github.com/digital-asset/daml/blob/main/navigator/backend/src/test/scala/com/digitalasset/navigator/backend/query/ContractSortSpec.scala), but the test coverage is very bad. In particular, it does not cover the `checkParameter` method that was modified by this PR.

BTW, the entire custom query engine should (in my opinion) be replaced by SQL queries, so I'm not sure how much we should invest in improving the testing of the engine.

Reply:
Adding a test:

You can extend [ContractSortSpec.scala](https://github.com/digital-asset/daml/blob/main/navigator/backend/src/test/scala/com/digitalasset/navigator/backend/query/ContractSortSpec.scala) with the following lines at the bottom:

```
  test(List("template.parameter.foo" -> ASCENDING), List(contract2, contract4, contract1, contract3))
  test(List("template.parameter.int" -> ASCENDING), List(contract1, contract3, contract2, contract4))
```

The first line sorts the test contracts by `template.parameter.foo`, i.e., by the `foo` field of the contract parameter. That field has value `text` for the contracts 1 and 3, and does not exist for contracts 2 and 4.

You could then extend the test coverage by changing [damlLfRecord0](https://github.com/digital-asset/daml/blob/main/navigator/backend/src/test/scala/com/digitalasset/navigator/backend/query/ContractSortSpec.scala#L32) and [damlLfRecord1](https://github.com/digital-asset/daml/blob/main/navigator/backend/src/test/scala/com/digitalasset/navigator/backend/query/ContractSortSpec.scala#L39) to include other data types, and adding additional test cases for them.